Add GCE authority plugin

matt0bi opened the pull request:
**What this PR does / why we need it**:

This PR adds the `GCEAuthority` plugin, which first fetches and filters instance data from GCE, then creates messages to send to the reconciler plugin.

**Which issue(s) this PR fixes**
<!-- optional; in `fixes #<issue number>, fixes #<issue_number>, ...` format, will close the issue(s) when PR gets merged: -->
None.

**Special notes for your reviewer**:

This set of changes includes the feedback that I received on #7. One item that I think we should discuss in person is what zone information the reconciler plugin expects and how this will affect instance filtering.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

@spotify/alf

matt0bi committed:
Add plugins.authority module

matt0bi commented on docs/config.rst:
```diff
@@ -39,7 +39,7 @@ Any configuration key/value listed here may also be used in the specific plugin
     To learn more about GCP projects, please see `Google's docs on creating & managing projects <https://cloud.google.com/resource-manager/docs/creating-managing-projects>`_.
 
-.. option:: scopes=["STR", "STR"]
+.. option:: scopes=["STR","STR"]
```
Sphinx was complaining about this when other options specified the same option (eg. `crm_scopes=["STR", "STR"]`). When the space is removed, other `option` statements can specify the same type attributes.

econchick replied:
so weird.

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+from gordon_janitor_gcp import msg_generators
+
+
+class GCEAuthorityBuilder:
```
Hmm, I'm starting to dislike a lot of this divergence of style... I'd like to talk about this Monday AM.

matt0bi replied:
Sure! Just lemme know whenever's good.

matt0bi replied:
Talked about this in person :tada:

econchick commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,70 @@
+    def __init__(self, zone):
+        self.zone = zone
+
+    async def generate_msg(self, instance):
```
Looking at these two methods, do these need to be async? (also asked on the [PR that adds the interface definition](https://github.com/spotify/gordon-janitor/pull/6)).

econchick commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,70 @@
+            dict: instance data formatted for submitting to a channel.
+        """
+        return {
+            'name': instance.get('hostname'),
```
Do you want a message to be created if no `'hostname'` or `'internal_ip'` is present?

matt0bi replied:
Talked IRL about this: at this point in execution, the instance dict must have `hostname` and `external_ip` keys.

econchick commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,70 @@
+        return {
+            'name': instance.get('hostname'),
+            'type': 'A',
+            'rrdatas': [instance.get('internal_ip')]
```
Wouldn't the common use case be for external IP instead of internal? I don't know if internal is more of a spotify thing (it may not be), but I feel like there'd definitely be a common need for external.

econchick commented on tests/unit/test_msg_generators.py:
```diff
@@ -0,0 +1,62 @@
+        'rrdatas': [instance_data['internal_ip']],
+        'type': 'A'
+    }
+    assert result == expected
```
Don't forget, assert expected versus actual.

econchick reviewed (commented):
This is blending my thoughts for [this PR](https://github.com/spotify/gordon-janitor/pull/6) and here: I'm wondering if there even should be a message generator interface. It seems more plugin-specific that we do `{'zone': 'zone-name', 'rrsets': [{...}, ...]}` since that's how we've defined an event message for the GCP plugin for Gordon core. So in general I question the need for an interface definition like this.

Half-baked thought: could it be more useful to separate out the message type from the generator? i.e. have an `ARecordMessage`, `CNAMERecordMessage`, maybe even `PTRRecordMessage`. Then for each we can set up validation, i.e. max length for an A or CNAME record's name, no more than one rrset for a CNAME, etc.

I think a generator function (or class if need be) is fine to have, but I'm not convinced that 1) it's useful to have for all plugins (may be unnecessarily restrictive, and/or unnecessarily optimizing), and 2) that it needs to generate a specific type of record message.

econchick commented on docs/config.rst:
```diff
@@ -65,3 +65,41 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+
+.. option:: gce_keyfile="STR"
+
+    `Required`: Path to a Service Account JSON file, which will be used for authenticating with the Google Cloud Engine API. *NOTE*: The same SA account can be used for the `crm_keyfile` setting.
```
Could someone just have `keyfile` defined to use for both?

econchick commented on docs/config.rst:
```diff
@@ -65,3 +65,41 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+
+.. option:: project_blacklist=["STR","STR"]
+
+    `Optional`: List of unique, user-assigned project IDs (`projectId`) that will be ignored when fetching projects.
```
Did you want `projectId` to be italics, or code block?

econchick commented on docs/config.rst:
```diff
@@ -65,3 +65,41 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
     For the Pub/Sub plugin, ``keyfile`` is not required when running against the `Pub/Sub Emulator <https://cloud.google.com/pubsub/docs/emulator>`_ that Google provides.
+
+
+gcp.gceauthority
```
We need to decide on naming. I have `gcp.gdns` for the reconciler and `gcp.gpubsub` for the publisher.

Maybe we should do something like `gcp.authority`, `gcp.reconciler`, and `gcp.publisher`. We can then nest config, like so:

```toml
[gcp]
# general config
[gcp.authority]
zone = "foo"
keyfile = "/path/to/shared/gce-gcrm/keyfile.json"
[gcp.authority.gce]
keyfile = "/path/to/gce/keyfile.json"  # if gce & gcrm weren't sharing the keyfile
scopes = ["one-scope", "another-scope"]
[gcp.authority.gcrm]
keyfile = "/path/to/gcrm/keyfile.json"
scopes = ["yet-another-scope"]
[gcp.reconciler]
# ...
```

This would also lend to flexibility of specific GCP products - i.e. if someone didn't want GCE to be the authority, but maybe GKE (???) or a GCP storage product, or something.

If we go down this route, we'd have to take a look at how plugin config parsing + plugin initialization behaves.

matt0bi replied:
I simplified the config section to the minimum in order to keep it generic.

econchick commented on docs/config.rst:
```diff
@@ -65,3 +65,41 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+
+Additional plugin-specific configuration is needed:
+
+.. option:: zone="STR"
```
This tells me we have to run a janitor per zone. Is there not a way to support multiple?

matt0bi replied:
Talked about this IRL, will use single zone in order to keep things generic for other users.

econchick commented on gordon-janitor.toml.example:
```diff
@@ -14,3 +14,13 @@ scopes = ["ndev.clouddns.readonly"]
 topic = "dns-changes-topic"
+
+[gcp.gceauthority]
+crm_keyfile = "/path/to/crm-service-account.json",
```
TOML lines don't end with a comma.

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+
+    import asyncio
+
+    import gce_authority
```
absolute path plz?

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+
+    async def done(self):
+        """Clean up after a run."""
+        await self.rrset_channel.put(None)
```
You'll need to also close the session(s) for the clients that this authority uses.

econchick replied:
Also, here would be where the plugin would need to cleanup its tasks it created if we choose to go that route.

econchick replied:
Ah and a log line saying the authority is complete or something.

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+            yield await self.gce_client.list_instances(project)
+
+    async def start(self):
+        """Gather instance data batches and send them to the
```
This needs to be one single line [per the style guide](https://google.github.io/styleguide/pyguide.html#Comments).

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+
+    async def _get_instances(self, projects):
+        for project in projects:
+            yield await self.gce_client.list_instances(project)
```
Another half-baked idea: maybe in the GCEClient, this method doesn't actually [filter out for the zone](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/gcp_clients.py#L179), it just gets all instances. Then have the authority be responsible for sorting into zone(s) per config? It seems like the GCEClient should just worry about getting instances for a project given, and then the authority responsible for what it cares about.

matt0bi replied:
I'd rather this go in the TODO doc for discussion since it isn't part of the scope of this PR.

econchick commented on tests/unit/test_gce_authority.py:
```diff
@@ -0,0 +1,188 @@
+    kwargs = {'other': 'kwargs'}
+    builder = gce_authority.GCEAuthorityBuilder(
+        authority_config, rrset_channel, **kwargs)
+    session = builder.session
```
Do you not need to yield for this? I'd expect aiohttp to complain about a session not being closed.

matt0bi replied:
I'm just "grabbing" the attribute of the builder and assigning it to its own variable for clarity so I don't have to yield anything. The next line (`with contextlib.closing(session))`) takes care of closing the session.

econchick replied:
hm wonder how I missed that...

econchick commented on tests/unit/test_gce_authority.py:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    # one message per each project
+    assert (await rrset_channel.get()) == expected_msg
```
another one that's backwards re: expected vs actual :)

econchick commented on tests/unit/test_gce_authority.py:
```diff
@@ -0,0 +1,188 @@
+    target = 'gordon_janitor_gcp.gce_authority.GCEAuthorityBuilder'
+    authority_obj = mocker.Mock()
+    authority_class = mocker.MagicMock(
+        return_value=authority_obj)
```
is there a reason for the line break?

econchick commented on tests/unit/test_gce_authority.py:
```diff
@@ -0,0 +1,188 @@
+async def test_builder_error_on_required_config(authority_config, caplog):
+    caplog.set_level(logging.ERROR)
+    del authority_config['crm_keyfile']
+    rrset_channel = 'channel'
```
Is there a reason why you create a proper queue elsewhere, but just a string here?

econchick commented on tests/unit/test_gce_authority.py:
```diff
@@ -0,0 +1,188 @@
+    authority_obj = mocker.Mock()
+    authority_class = mocker.MagicMock(
+        return_value=authority_obj)
+    mocker.patch(target, new=authority_class)
```
Actually, I'm not sure I understand the whole need to patch this if you're calling `get_authority.get_authority` directly?

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+                 gce_client,
+                 msg_generator,
+                 rrset_channel=None,
+                 **kwargs):
```
I really _really_ dislike this style :-/ It's not stated in the google python style guide, and a little elusive in pep8, but we should be consistent. It's also weird to read it this way when there is one line per variable definition right after. Maybe @descent3me can weigh in.

econchick commented on gordon_janitor_gcp/gce_authority.py:
```diff
@@ -0,0 +1,181 @@
+        async for instances in self._get_instances(projects):
+            rrsets = await self.msg_generator.generate_batch(instances)
+            await self.rrset_channel.put(rrsets)
+        # TODO: emit a metric of domain records created per zone.
```
maybe records per project too?

econchick reviewed (changes requested):
Could you also change the title of this PR? I've accidentally done this before too - just left it as the branch name 😬

econchick commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,70 @@
+    >>> loop = asyncio.get_event_loop()
+    >>> instance1_data = {'hostname': 'a-host', 'internal_ip': '192.168.11.5'}
+    >>> instance2_data = {'hostname': 'b-host', 'internal_ip': '192.168.11.6'}
+    >>> generator = ARecordMsgGenerator('example.com')
```
Prob need to import this first :)

matt0bi commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+
+        instances = []
+        async for project_instances in self._get_instances(projects):
+            instances.extend(project_instances)
```
@econchick I used `extend` here because it's a bit clearer. I also looked into what you said, about `extend` being less efficient, but I couldn't find anything about this except the Python [TimeComplexity - list](https://wiki.python.org/moin/TimeComplexity) page, which lists that single appends are O(1) whereas extends are O(k), which makes it sound like they should be similarly efficient.

matt0bi commented:
Hi @econchick, I addressed your feedback, could you please re-review?

- Went with the "one zone + filter ability" idea we talked about yesterday.
- Got rid of the message generator idea. I think it was a good idea, but I'll wait and see if it could be handier in the future.
- Moved the code around in order to conform to the new shape of the project.

If I omitted to address anything, I'm super sorry!

econchick commented on docs/config.rst:
```diff
@@ -67,4 +67,31 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+
+.. option:: tag_blacklist=["STR","STR"]
+
+    `Optional`: List of network tags that will be used to filter out unwanted GCE instances.
```
It might be handy to link to [network tags](https://cloud.google.com/vpc/docs/add-remove-network-tags) and [metadata](https://cloud.google.com/compute/docs/storing-retrieving-metadata).

matt0bi replied:
Good idea, thanks for providing the links, too!

econchick commented on docs/config.rst:
```diff
@@ -67,4 +67,31 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+
+.. option:: instance_filter="STR"
+
+    `Optional`: String used to filter instances by instance attributes. It is passed directly to GCE's v1.instances.aggregatedList endpoint's `filter` parameter.
```
Maybe link to filtering here as well.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+    def _get_or_raise(self, key):
+        item = self.config.get(key)
+        if not item:
+            msg = f'gcp.gce configuration missing "{key}" setting'
```
Hmm I think I'd prefer to have more helpful messaging consistent with the rest of the plugins ([reconciler keyfile](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/plugins/reconciler.py#L93-L94), [reconciler project](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/plugins/reconciler.py#L98), [publisher keyfile](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/plugins/publisher.py#L94-L95), etc). I know that might change the logic of this helper method a little bit.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+            msg = f'gcp.gce configuration missing "{key}" setting'
+            logging.error(msg)
+            # Close the session if any required config is missing.
+            self.session.close()
```
Do we have to create a session before we use it, to avoid needing to close it here?

matt0bi replied:
We could let the clients create their own sessions, but that would increase the number of sessions from 1 to 2 here, and to 3 in the internal-facing parts. I'm going for minimizing the number of moving parts - do you think that's too cautious?

econchick replied:
I'm thinking perhaps to have `self.session` set to `None` on instantiation, and then this:

```py
def build_authority(self):
    self._get_or_raise('zone')
    keyfile_path = self._get_or_raise('keyfile')
    scopes = self._get_or_raise('scopes') 
    self.session = aiohttp.ClientSession()
    crm_client = self.get_crm_client(keyfile_path, scopes)
    # and so on
```

econchick commented on docs/config.rst:
```diff
@@ -67,4 +67,31 @@ All configuration options above in the general ``[gcp]`` may be used here. Addit
+gcp.gce
+~~~~~~~
+
+All configuration options from the general ``[gcp]`` section may be used here.
```
So there's no config that would be different between GCE and GCRM APIs? Is it a conscious decision to make folks use one keyfile for both APIs?

matt0bi replied:
This is the "simple defaults" approach because both clients can share the same keyfile and scopes and work fine and I think the multi-keyfile approach could be premature at this point.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+
+
+class GCEAuthorityBuilder:
+    """Class that helps create a properly configured GCEAuthority object.
```
Should be in the imperative, like "Build and configure a :class:`GCEAuthority` object".

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+                             blacklisted_metadata=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
```
"Coordinate" is weird, lol. Setup?

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
```
Also should be in the imperative, e.g. "Get instance data from source of truth."

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+        }
+
+    async def start(self):
+        """Batch instance data and send it to the :py:obj:`self.rrset_channel`.
```
You can drop the `:py` prefix.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
```
You can use the actual types here, e.g. `gce.GCRMClient`

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+        project_blacklist = set(self.config.get('project_blacklist', []))
+        # TODO: emit a metric for all projects and a metric for
+        # project - blacklist.
+        return sorted(projects - project_blacklist)
```
OK I'm only have done with my coffee, so pardon me for the silly q: why sorted?

matt0bi replied:
Just to introduce determinism into the mix - having the projects always be processed in alphabetical order can help reasoning about/debugging. I can remove it if you think it's confusing?

econchick commented on setup.py:
```diff
@@ -97,6 +97,7 @@ def install_requires():
         'gordon_janitor.plugins': [
             'gcp.gdns = gordon_janitor_gcp:get_reconciler',
             'gcp.gpubsub = gordon_janitor_gcp:get_publisher',
+            'gcp.gce = gordon_janitor:get_authority',
```
ABC order?

econchick commented on tests/unit/plugins/test_authority.py:
```diff
@@ -0,0 +1,148 @@
+    builder = authority.GCEAuthorityBuilder(
+        authority_config, rrset_channel, **kwargs)
+    session = builder.session
+    with pytest.raises(exceptions.GCPConfigError), contextlib.closing(session):
```
May be a bit pedantic, but perhaps match at least part of the error message that's raised?

matt0bi replied:
I removed the test since the same behavior was being tested in `test_init.py::test_get_authority_config_raises`.

econchick commented on tests/unit/plugins/test_authority.py:
```diff
@@ -0,0 +1,148 @@
+        'hostname': f'host-{i}',
+        'internal_ip': f'192.168.1.{i}',
+        'external_ip': f'1.1.1.{i}'
+    } for i in range(1, 4)]
```
This is a little confusing to read, would you mind doing something like this:

```py
_data = {
    'hostname': 'host-{i}',
    'internal_ip': '192.168.1.{i}',
    'external_ip': '1.1.1.{i}'
}
instance_data = [_data.format(i=i) for i in range(1 ,4)]
```

matt0bi replied:
Thanks for pointing this out. I wasn't able to think of a way to keep the comprehension and format the data, so I exploded it out into a for-loop, which looks much clearer.

econchick commented on tests/unit/plugins/test_authority.py:
```diff
@@ -0,0 +1,148 @@
+        'name': record['hostname'],
+        'type': 'A',
+        'rrdatas': [record['external_ip']]
+    } for record in instance_data] * 2
```
This one's a little hard to read, too. Not sure if there's an approach to improve this, so if not, then nvm.

matt0bi replied:
Exploded this one into a for-loop as well.

econchick commented on tests/unit/plugins/test_authority.py:
```diff
@@ -0,0 +1,148 @@
+    # one message includes data for all projects
+    assert expected_msg == (await rrset_channel.get())
+    # start also calls self.done at the end
+    assert None is (await rrset_channel.get())
```
`assert not (await rrset_channel.get())`?

matt0bi replied:
I'd prefer to check specifically for `None` since the plugins check for `None` and not for falsy things, too:

```python
# reconciler.py
desired_rrset = await self.rrset_channel.get()
if desired_rrset is None:
```

econchick reviewed (commented):
Just some small things - nice job!

matt0bi commented:
Hey @econchick, I addressed your latest feedback, could PTAL when you have a chance?

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+        """
+        self._get_or_raise('zone')
+        keyfile_path = self._get_or_raise('keyfile')
+        scopes = self._get_or_raise('scopes')
```
Do scopes need to be here? Isn't there a default we can use?

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,185 @@
+        self.rrset_channel = rrset_channel
+        self.kwargs = kwargs
+        self._loop = asyncio.get_event_loop()
+        self.session = aiohttp.ClientSession(loop=self._loop)
```
I think `aiohttp` is already doing the `asyncio.get_event_loop` under the hood, so we don't need to pass it in. We can, however, accept an explicit loop argument to make it <3.6 compatible, and do something like `self._loop = loop or asyncio.get_event_loop()`.

matt0bi replied:
Thanks, I was thinking of some earlier version of aiohttp that would complain or error out if it was created outside of an async context.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,192 @@
+        """Set up dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
```
I think you can actually refer to the actual object name, instead of "obj:" it can be "\`GCEAuthority\`:"

matt0bi replied:
Thanks for the tip - I looked at the other builder's `build_*` method and noticed they don't have a docstring at all, so I removed the whole thing for consistency.

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,192 @@
+        try:
+            self._validate_config()
+        except exceptions.GCPConfigError:
+            self.session.close()
```
It may have been lost in your most recent push, but I think setting the `self.session` object after you validate config would be better. (I also don't think you're properly re-raising the config error) i.e:

```py
class GCEAuthorityBuilder:
    def __init__(...):
        self.session = None

    def build_authority(self):
        self._validate_config()  # will raise on its own, no need to catch + reraise
        self.session = aiohttp.ClientSession()
        # continue on
```

econchick commented on gordon_janitor_gcp/plugins/authority.py:
```diff
@@ -0,0 +1,192 @@
+        msg = 'Finished sending record messages to the reconciler.'
+        logging.info(msg)
+        await self.rrset_channel.put(None)
+        self.gce_client._session.close()
```
Does this need an `await`?

econchick reviewed (commented):
Just a few more thoughts.

matt0bi commented:
Hey @econchick, thanks for reviewing again - I addressed your comments, mind having another look when you have the time?

econchick reviewed (approved):
I think this looks good! 🎉 🌮

matt0bi commented:
Woo! Thanks @econchick :tada:
